lib/puppet/pops/types/type_calculator.rb

```diff
@@ -1192,22 +1201,34 @@ def tuple_entry_at(tuple_t, from, to, index)
   # @api private
   #
   def assignable_PStructType(t, t2)
-    return true if t == t2 || t.elements.empty? && (t2.is_a?(Types::PHashType))
-    h = t.hashed_elements
+    els = t.elements
+    return true if t == t2 || els.empty? && (t2.is_a?(Types::PHashType))
     if t2.is_a?(Types::PStructType)
       h2 = t2.hashed_elements
-      h.size == h2.size && h.all? {|k, v| assignable?(v, h2[k]) }
```

Comment:
this line should be changed to subtract keys, and then check assignability of either the corresponding type in the t2 struct, or PNilType if that key did not have a type.